src/schedule/dma_multi_chan_domain.c

```diff
@@ -0,0 +1,259 @@
+}
+
+static void dma_multi_chan_domain_unregister(struct ll_schedule_domain *domain,
+					     uint32_t num_tasks)
+{
+	struct dma_domain *dma_domain = ll_sch_domain_get_pdata(domain);
+	struct dma *dmas = dma_domain->dma_array;
+	int core = cpu_get_id();
+	int i;
+	int j;
+
+	for (i = 0; i < dma_domain->num_dma; ++i) {
+		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+			/* channel not set as scheduling source */
+			if (!dma_is_scheduling_source(&dmas[i].chan[j]))
+				continue;
+
+			/* channel still running */
+			if (dmas[i].chan[j].status == COMP_STATE_ACTIVE)
+				continue;
+
+			/* channel owned by different core */
+			if (core != dmas[i].chan[j].core)
+				continue;
+
+			/* channel hasn't been running */
+			if (!(dma_domain->channel_mask[i][core] & BIT(j)))
+				continue;
+
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_MASK);
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_CLEAR);
+			interrupt_clear_mask(dma_domain->data[i][j].irq,
+					     BIT(j));
+
+			dma_domain->data[i][j].task = NULL;
+			dma_domain->channel_mask[i][core] &= ~BIT(j);
+
+			/* unregister interrupt */
+			if (!dma_domain->aggregated_irq)
+				dma_multi_chan_domain_irq_unregister(
+						&dma_domain->data[i][j]);
+			else if (!dma_domain->channel_mask[i][core])
+				dma_multi_chan_domain_irq_unregister(
+						dma_domain->arg[i][core]);
+
+			return;
+		}
+	}
+}
+
+static bool dma_multi_chan_domain_is_pending(struct ll_schedule_domain *domain,
+					     struct task *task)
+{
+	struct dma_domain *dma_domain = ll_sch_domain_get_pdata(domain);
+	struct dma *dmas = dma_domain->dma_array;
+	uint32_t status;
+	int i;
+	int j;
+
+	for (i = 0; i < dma_domain->num_dma; ++i) {
+		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+			status = dma_interrupt(&dmas[i].chan[j],
+					       DMA_IRQ_STATUS_GET);
+			if (!status)
+				continue;
+
+			/* not this task */
+			if (dma_domain->data[i][j].task != task)
+				continue;
+
+			/* clear interrupt */
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_CLEAR);
+			interrupt_clear_mask(dma_domain->data[i][j].irq,
+					     BIT(j));
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+struct ll_schedule_domain *dma_multi_chan_domain_init(struct dma *dma_array,
+						      uint32_t num_dma, int clk,
+						      bool aggregated_irq)
+{
+	struct ll_schedule_domain *domain;
+	struct dma_domain *dma_domain;
+	struct dma *dma;
+	int i;
+	int j;
+
+	domain = domain_init(SOF_SCHEDULE_LL_DMA, clk,
+			     &dma_multi_chan_domain_ops);
+
+	dma_domain = rzalloc(RZONE_SYS | RZONE_FLAG_UNCACHED, SOF_MEM_CAPS_RAM,
+			     sizeof(*dma_domain));
+	dma_domain->dma_array = dma_array;
+	dma_domain->num_dma = num_dma;
+	dma_domain->aggregated_irq = aggregated_irq;
+
+	/* retrieve IRQ numbers for each DMA channel */
+	for (i = 0; i < num_dma; ++i) {
+		dma = &dma_array[i];
+		for (j = 0; j < dma->plat_data.channels; ++j)
+			dma_domain->data[i][j].irq = interrupt_get_irq(
+				dma_chan_irq(dma, j), dma_irq_name(dma));
```

Comment:
I suspect each individual platform should add its own definition of dma_chan_irq and that all IRQs for a particular controller map to the same interrupt controller (this is currently true but may become false in the future unless some rework is done in the interrupt system so I don't need to do the currently existing hack on the IRQ_STEER driver)

Finally understood!